File: Programmers/노수일/Q42579.py
def solution(genres, plays):
    dictionary = {}

    for i in range(len(genres)):
        genre = genres[i]
        play = plays[i]

        if genre in dictionary:
            dictionary[genre]["count"] += play
            dictionary[genre][i] = play
        else:
            dictionary[genre] = {"count": play, i: play}

    sorted_keys = sorted(dictionary.keys(), key=lambda x: dictionary[x]["count"], reverse=True)
    answer = []

    for key in sorted_keys:
        sorted_value_keys = sorted(dictionary[key].keys(), key=lambda x: dictionary[key][x], reverse=True)
        end = 3
        if len(sorted_value_keys) == 2:
            end = 2
        for i in range(1, end):
            answer.append(sorted_value_keys[i])

    return answer


"""
{
    "classic" : {
        "count" : 1450,
        0 : 500,
        2 : 150,
        3 : 800
    },
    "pop" : {
        "count" : 3100,
        1 : 600,
        4 : 2500
    }
}
"""

print(solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500]))

File: Programmers/유석환/Q42579.swift
import Foundation

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var songInfoByGenre = [String: [(Int, Int)]]() // [장르이름: (고유번호, 재생횟수)]
    var playCountByGenre = [String: Int]() // [장르이름: 재생횟수 총합]
    
    for i in 0..<genres.count {
        if songInfoByGenre[genres[i]] != nil {
            songInfoByGenre[genres[i]]!.append((i, plays[i]))
        } else {
            songInfoByGenre[genres[i]] = [(i, plays[i])]
        }
        
        if playCountByGenre[genres[i]] != nil {
            playCountByGenre[genres[i]]! += plays[i]
        } else {
            playCountByGenre[genres[i]] = plays[i]
        }
    }
    
    for key in songInfoByGenre.keys {
        songInfoByGenre[key]!.sort(by: { $0.1 > $1.1 })
    }
    
    let sortedArr = playCountByGenre.sorted(by: { $0.value > $1.value })
    var result = [Int]()
    
    for item in sortedArr {
        result.append(songInfoByGenre[item.key]![0].0)
        
        if songInfoByGenre[item.key]!.count > 1 {
            result.append(songInfoByGenre[item.key]![1].0)
        }
    }
    
    return result
}
